eras/shelley/impl/src/Cardano/Ledger/Shelley/Scripts.hs

```diff
@@ -191,3 +195,29 @@ eqMultiSigRaw t1 t2 = go (getMemoRawType t1) (getMemoRawType t2)
     go (RequireAnyOf' ts1) (RequireAnyOf' ts2) = liftEq eqMultiSigRaw ts1 ts2
     go (RequireMOf' n1 ts1) (RequireMOf' n2 ts2) = n1 == n2 && liftEq eqMultiSigRaw ts1 ts2
     go _ _ = False
+
+-- | Script evaluator for native multi-signature scheme. 'vhks' is the set of
+-- key hashes that signed the transaction to be validated.
+evalMultiSig ::
+  Era era =>
+  Set.Set (KeyHash 'Witness (EraCrypto era)) ->
+  MultiSig era ->
+  Bool
+evalMultiSig vhks = go
+  where
+    -- The important part of this validator is that it will stop as soon as it reaches the
+    -- required number of valid scripts
+    isValidMOf n [] = n <= 0
+    isValidMOf n (msig : msigs) =
+      n <= 0 || if go msig then isValidMOf (n - 1) msigs else isValidMOf n msigs
+    go = \case
+      RequireSignature hk -> Set.member hk vhks
+      RequireAllOf msigs -> all go msigs
+      RequireAnyOf msigs -> any go msigs
+      RequireMOf m msigs -> isValidMOf m msigs
```

Comment:
How about 'evalMultiSig'  instead of go

Reply:
`evalMultiSig` is the name of the function. :smile: 

Are you talking about the local `isValidMOf`? This one only works for `MOf`, hence the name `isValidMOf`